File: app/db/admin_browser.py
"""Generic database-browser methods: schema/table listing, table data
preview, and truncation. Powers the panel's Database admin view."""

import copy
import time
from datetime import datetime
from typing import Any

import aiomysql

from .helpers import (
    PANEL_SCHEMA_CACHE_TTL_SECONDS,
    PANEL_TABLE_CACHE_TTL_SECONDS,
    PANEL_TABLE_DATA_CACHE_MAX_ITEMS,
    PANEL_TABLE_DATA_CACHE_TTL_SECONDS,
    SYSTEM_SCHEMAS,
)
from .identifiers import _validate_identifier


class AdminBrowserMixin:
    async def _table_exists(self, schema: str, table: str) -> bool:
        schema = _validate_identifier(schema, "schema")
        table = _validate_identifier(table, "table")
        key = (schema, table)
        now = time.monotonic()
        cached = self._table_exists_cache.get(key)
        if cached and cached[0] > now:
            return bool(cached[1])
        row = await self._fetchone(
            """
            SELECT 1 AS table_exists
            FROM information_schema.tables
            WHERE table_schema = %s AND table_name = %s
            LIMIT 1
            """,
            (schema, table),
        )
        exists = bool(row)
        self._table_exists_cache[key] = (now + PANEL_TABLE_CACHE_TTL_SECONDS, exists)
        return exists

    async def ping(self) -> bool:
        try:
            row = await self._fetchone("SELECT 1 AS ok")
            return bool(row and row.get("ok") == 1)
        except Exception:
            return False

    async def list_schemas(self) -> list[str]:
        now = time.monotonic()
        if self._schema_cache and self._schema_cache[0] > now:
            return list(self._schema_cache[1])
        rows = await self._fetchall(
            """
            SELECT schema_name
            FROM information_schema.schemata
            WHERE schema_name NOT IN ('information_schema', 'mysql', 'performance_schema', 'sys')
            ORDER BY schema_name
            """
        )
        schemas = [row["schema_name"] for row in rows]
        self._schema_cache = (time.monotonic() + PANEL_SCHEMA_CACHE_TTL_SECONDS, list(schemas))
        return schemas

    async def list_tables(self, schema: str) -> list[dict[str, Any]]:
        schema = _validate_identifier(schema, "schema")
        now = time.monotonic()
        cached = self._tables_cache.get(schema)
        if cached and cached[0] > now:
            return copy.deepcopy(cached[1])
        if cached:
            self._tables_cache.pop(schema, None)
        rows = await self._fetchall(
            """
            SELECT table_name, table_rows
            FROM information_schema.tables
            WHERE table_schema = %s
            ORDER BY table_name
            """,
            (schema,),
        )
        tables = [
            {"table_name": row["table_name"], "estimated_rows": int(row["table_rows"] or 0)}
            for row in rows
        ]
        self._tables_cache[schema] = (time.monotonic() + PANEL_TABLE_CACHE_TTL_SECONDS, copy.deepcopy(tables))
        return tables

    async def truncate_table(self, schema: str, table: str) -> None:
        schema = _validate_identifier(schema, "schema")
        table = _validate_identifier(table, "table")
        if schema in SYSTEM_SCHEMAS:
            raise ValueError(f"Refusing operation on system schema: {schema}")
        pool = await self._get_pool()
        async with pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute("SET FOREIGN_KEY_CHECKS = 0")
                try:
                    await cur.execute(f"TRUNCATE TABLE `{schema}`.`{table}`")
                finally:
                    await cur.execute("SET FOREIGN_KEY_CHECKS = 1")
        self._invalidate_hot_caches()

    async def truncate_schema(self, schema: str) -> dict[str, Any]:
        schema = _validate_identifier(schema, "schema")
        if schema in SYSTEM_SCHEMAS:
            raise ValueError(f"Refusing operation on system schema: {schema}")
        pool = await self._get_pool()
        tables = await self.list_tables(schema)
        async with pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute("SET FOREIGN_KEY_CHECKS = 0")
                try:
                    for table in tables:
                        table_name = _validate_identifier(table["table_name"], "table")
                        await cur.execute(f"TRUNCATE TABLE `{schema}`.`{table_name}`")
                finally:
                    await cur.execute("SET FOREIGN_KEY_CHECKS = 1")
        self._invalidate_hot_caches()
        return {"schema": schema, "truncated_tables": len(tables), "tables": [t["table_name"] for t in tables]}

    async def get_table_data(self, schema: str, table: str, limit: int = 100) -> dict[str, Any]:
        schema = _validate_identifier(schema, "schema")
        table = _validate_identifier(table, "table")
        if schema in SYSTEM_SCHEMAS:
            raise ValueError(f"Refusing operation on system schema: {schema}")
        safe_limit = max(1, min(int(limit or 100), 500))
        cache_key = (schema, table, safe_limit)
        now = time.monotonic()
        cached = self._table_data_cache.get(cache_key)
        if cached and cached[0] > now:
            self._table_data_cache.move_to_end(cache_key)
            return copy.deepcopy(cached[1])
        if cached:
            self._table_data_cache.pop(cache_key, None)

        pool = await self._get_pool()

        async with pool.acquire() as conn:
            # Use DictCursor so the frontend gets column names alongside the values
            async with conn.cursor(aiomysql.DictCursor) as cur:
                await cur.execute(f"SELECT * FROM `{schema}`.`{table}` LIMIT %s", (safe_limit,))
                rows = await cur.fetchall()
                
                # Sanitize the data for JSON serialization
                processed_rows = []
                for row in rows:
                    processed_row = {}
                    for key, val in row.items():
                        if isinstance(val, datetime):
                            processed_row[key] = val.isoformat()
                        elif isinstance(val, bytes):
                            processed_row[key] = "<binary data>"
                        else:
                            processed_row[key] = val
                    processed_rows.append(processed_row)

                result = {
                    "schema": schema,
                    "table": table,
                    "count": len(processed_rows),
                    "rows": processed_rows
                }
                self._table_data_cache[cache_key] = (time.monotonic() + PANEL_TABLE_DATA_CACHE_TTL_SECONDS, copy.deepcopy(result))
                self._table_data_cache.move_to_end(cache_key)
                while len(self._table_data_cache) > PANEL_TABLE_DATA_CACHE_MAX_ITEMS:
                    self._table_data_cache.popitem(last=False)
                return result

    def _json_value(self, value: Any) -> Any:
        if hasattr(value, "isoformat"):
            return value.isoformat()
        if isinstance(value, bytes):
            return "<binary data>"
        return value

    def _json_row(self, row: dict[str, Any]) -> dict[str, Any]:
        return {key: self._json_value(value) for key, value in row.items()}

File: lua/src/admin_browser.lua
-- Generic database-browser methods: schema/table listing, table data
-- preview, and truncation. Port of app/db/admin_browser.py.
--
-- SQL-INJECTION NOTE (this is the part priority item 7 in the task
-- specifically flagged to watch for): every schema/table name that reaches
-- raw SQL text (not a %s-bound parameter — Postgres doesn't allow binding
-- identifiers as parameters at all) is run through validate_identifier()
-- first, an allowlist regex (`^[A-Za-z0-9_]+$`) ported byte-for-byte from
-- app/db/identifiers.py. That regex rejects quotes, dots, whitespace,
-- semicolons, comment markers — anything that could break out of the
-- identifier position. Table/column names for `SELECT * FROM <schema>.<table>`
-- and `UPDATE <table> SET <col> = %s` all go through this gate.
local db = require("db")
local config = require("config")

local M = {}

local IDENTIFIER_PATTERN = "^[%w_]+$"

function M.validate_identifier(value, field_name)
  value = tostring(value or ""):match("^%s*(.-)%s*$")
  if value == "" or not value:match(IDENTIFIER_PATTERN) then
    error("Invalid " .. (field_name or "identifier") .. ": " .. tostring(value), 0)
  end
  return value
end

-- "Schema" in the Python/MariaDB original means "one of the MySQL schemas
-- visible from a single connection". This Postgres port keeps one
-- connection PER DATABASE (see db.lua's module doc comment), so the
-- equivalent concept is "one of the swarm's known database names" — the
-- fixed set from config.lua, not a live `pg_database` scan (which would also
-- surface unrelated databases on the same Postgres server, e.g.
-- `image_upscaler`/`postgres`/`template0`, that aren't part of this panel's
-- domain and shouldn't be truncatable through this admin UI).
function M.list_schemas()
  local seen, out = {}, {}
  local function add(name)
    if name and name ~= "" and not seen[name] then
      seen[name] = true
      out[#out + 1] = name
    end
  end
  for _, bot in ipairs(config.all_bots or {}) do add(bot.db_schema) end
  add("accountlogins")
  add("discord_aria")
  add("image_gallery")
  table.sort(out)
  return out
end

function M.list_tables(schema)
  schema = M.validate_identifier(schema, "schema")
  local rows = db.fetchall(
    schema,
    [[SELECT c.relname AS table_name,
             COALESCE(c.reltuples, 0)::bigint AS estimated_rows
      FROM pg_class c
      JOIN pg_namespace n ON n.oid = c.relnamespace
      WHERE n.nspname = 'public' AND c.relkind = 'r'
      ORDER BY c.relname]]
  )
  local out = {}
  for _, row in ipairs(rows) do
    -- Postgres sets pg_class.reltuples = -1 for a table that has never been
    -- ANALYZE'd (fresh/empty tables) — clamp to 0 rather than surfacing a
    -- confusing negative "row count" in the admin UI.
    out[#out + 1] = { table_name = row.table_name, estimated_rows = math.max(0, db.toint(row.estimated_rows, 0)) }
  end
  return out
end

local function json_value(v)
  return v -- pg.lua already returns text/bool/number in JSON-safe Lua forms; no bytea/datetime special-casing needed here (see get_table_data below for the one case that does need it: bytea).
end
M.json_row = function(row)
  local out = {}
  for k, v in pairs(row or {}) do out[k] = json_value(v) end
  return out
end

function M.get_table_data(schema, table_name, limit)
  schema = M.validate_identifier(schema, "schema")
  table_name = M.validate_identifier(table_name, "table")
  local safe_limit = math.max(1, math.min(tonumber(limit) or 100, 500))
  local rows = db.fetchall(schema, string.format('SELECT * FROM "%s" LIMIT %%s', table_name), safe_limit)
  local processed = {}
  for _, row in ipairs(rows) do
    local out = {}
    for k, v in pairs(row) do
      -- pgmoon returns bytea as a raw Lua string; there's no reliable way to
      -- tell "binary" from "text" apart post-hoc, so unlike the Python
      -- original (which gets a typed `bytes` object from aiomysql and can
      -- check `isinstance(val, bytes)`) this just passes strings through as-is.
      -- Not currently an issue for any table this browser is pointed at
      -- (no bytea columns in accountlogins/image_gallery/discord_* schemas).
      out[k] = v
    end
    processed[#processed + 1] = out
  end
  return { schema = schema, table = table_name, count = #processed, rows = processed }
end

function M.truncate_table(schema, table_name)
  schema = M.validate_identifier(schema, "schema")
  table_name = M.validate_identifier(table_name, "table")
  db.execute(schema, string.format('TRUNCATE TABLE "%s" CASCADE', table_name))
end

function M.truncate_schema(schema)
  schema = M.validate_identifier(schema, "schema")
  local tables = M.list_tables(schema)
  for _, t in ipairs(tables) do
    db.execute(schema, string.format('TRUNCATE TABLE "%s" CASCADE', M.validate_identifier(t.table_name, "table")))
  end
  local names = {}
  for _, t in ipairs(tables) do names[#names + 1] = t.table_name end
  return { schema = schema, truncated_tables = #tables, tables = names }
end

return M
